### PR TITLE
Update Fetch Lesson

### PR DESCRIPTION
**Pull request by coreyladovsky**
This update adds fetch behavior with request objects, removes code that doesn't teach fetch, and adds a challenge section that can be done during lecture.

### CODE REVIEW

File: html_css_dom/fetch/README.md [hunk 1]
```diff
@@ -56,7 +56,7 @@ doSomething(result => {
 
 
 ```
-With modern functions, we attach our callbacks to the returned promises instead. Promises return a promise. Adding a `.then()` creates a promise chain:
+With modern functions, we attach our callbacks to the returned promises instead. __Promises return a promise.__ Adding a `.then()` creates a promise chain:
 
 ```js
 doSomething().then(result => {
```
Comment (mikeboyle) [inline html_css_dom/fetch/README.md:62]: Question here is actually about the next paragraph about the code block: https://github.com/joinpursuit/Pursuit-Core-Web/pull/136/files#diff-e64d184dfc27be0878a9dd8ea59a634c1f96640e6e28e9b62a9758f3be790e23R75
```
The `.then` takes in two optional arguments. The first argument is a function that will be fired upon the promise succeeding, the second argument is the function that will be fired upon the promise failing.
```
Do we need to spell out that the callback to `then` itself takes an argument (called `result` in the code block), and that `result` is whatever the previous promise returns?

Might also be good to point out that if you are going to chain together promises with `.then`, your callbacks will all probably need to return something to the next `.then()`. The result of the original promise is not in scope of the second, third, fourth `then()`. I'm guessing this is easy to overlook and is the source of a lot of bugs.

File: html_css_dom/fetch/README.md [hunk 2]
```diff
@@ -81,7 +81,7 @@ The arguments to `then` are option and  `catch(failureCallback)` is short for `t
 Although you will most likely already be dealing with asynchronous calls that return promises, it is possible to create your own promise with its constructor function. To do this you write new Promise and then pass in a callback function. That callback function will take in two arguments (each of which will be a function), one for resolve, and one for reject. Here's an example of what that might look like:
```
Comment (mikeboyle) [inline html_css_dom/fetch/README.md:81]: `one for resolve, and one for reject` ... do we need to spell that out more explicitly? For example the MDN docs say:
```
We call resolve(...) when what we were doing asynchronously was successful, and reject(...) when it failed.
```
I imagine it is also very mysterious to fellows what `resolve` and `reject` actually are, like:
- what exactly passes those callbacks to the promise? where do the `resolve` and `reject` functions come from and what do they do?
- do we have to use the parameter names `resolve` and `reject`? Would `foo` and `bar` work instead?

Is it necessary or feasible to preempt some of these questions in the README?

File: html_css_dom/fetch/README.md [hunk 3]
```diff
@@ -81,7 +81,7 @@ The arguments to `then` are option and  `catch(failureCallback)` is short for `t
...
 
 ```js
-let promise = new Promise(function(resolve, reject) {
+const promise = new Promise(function(resolve, reject) {
```
Comment (mikeboyle) [inline html_css_dom/fetch/README.md:84]: OK that the callback is not an ES6 arrow function?

File: html_css_dom/fetch/README.md [hunk 4]
```diff
@@ -172,172 +179,106 @@ Now, we'll need to add our event listeners in our `index.js`:
...
 
-    let flagImage = document.createElement('img')
-    flagImage.src = `${country.flag}`
-    flagImage.style = 'width:200px'
+What happens when you search for the number 23? 
 
-    newCard.appendChild(flagImage)
+You'll notice that an error is thrown. The error says that `countries.forEach is not a function`. This is a little bit unexpected and undesired. I'd like for you to add a `debugger` on the line before `return response.json()` and one before the `forEach` then try again. While you're paused on the first debugger take a look at the Response object. Some things you may notice is that the Response object has the properties `status` and `ok`. Notice that the status is 404 (Not Found) and ok is `false`. Now play through the debugger. Now our code is frozen on the line before the `forEach`. This is undesired. We had a 404 status code but we still fulfilled our promise and moved on to the next chain. This is a NEGATIVE of `fetch`. 4xx and 5xx status codes will NOT throw errors with `fetch`. Instead we need to alter our fetch code like so:
```
Comment (mikeboyle) [inline html_css_dom/fetch/README.md:205]: 👏 👏 👏  I love this, thank you for modeling debugging practices and helping fellows learn more about `fetch` through working with the code.

**Review by mikeboyle — APPROVED** [review]
LGTM! I have a couple of questions but nothing that should prevent you from merging this. Thank you also for creating and incorporating the playground backend! 🚀

File: html_css_dom/fetch/README.md (hunk 1 above)
Comment (coreyladovsky) [reply html_css_dom/fetch/README.md:62]: I think those suggestions make a lot of sense. I will update the code to include those.

File: html_css_dom/fetch/README.md (hunk 2 above)
Comment (coreyladovsky) [reply html_css_dom/fetch/README.md:81]: We definitely can talk more about those. I'm not sure if it's needed, b/c the fellows won't really be practicing building these promises.

File: html_css_dom/fetch/README.md (hunk 3 above)
Comment (coreyladovsky) [reply html_css_dom/fetch/README.md:84]: You're right, should probably be ES6

File: html_css_dom/fetch/README.md (hunk 2 above)
Comment (mikeboyle) [reply html_css_dom/fetch/README.md:81]: Fair enough, let's not go down the rabbit hole then.

**Pull request merged**